## Convert files from Thermo

ThermoRawFileParser https://github.com/compomics/ThermoRawFileParser

Convert profile MS1 data from Thermo raw files to centroid (MS1 and MS2) mzml

In [1]:
!mono resources/ThermoRawFileParser/ThermoRawFileParser.exe -d="data/raw/" -o="results/"

2021-10-25 16:04:26 INFO Started analyzing folder data/raw/
2021-10-25 16:04:26 INFO The folder contains 3 RAW files
2021-10-25 16:04:26 INFO Started parsing data/raw/Epemicins.raw
2021-10-25 16:04:27 INFO Processing 4611 MS scans
10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 

2021-10-25 16:04:36 INFO Finished parsing data/raw/Epemicins.raw
2021-10-25 16:04:36 INFO Started parsing data/raw/GermicidinB.raw
2021-10-25 16:04:37 INFO Processing 3619 MS scans
10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 

2021-10-25 16:04:42 INFO Finished parsing data/raw/GermicidinB.raw
2021-10-25 16:04:42 INFO Started parsing data/raw/GermicidinA.raw
2021-10-25 16:04:42 INFO Processing 3625 MS scans
10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 

2021-10-25 16:04:47 INFO Finished parsing data/raw/GermicidinA.raw


#### Copy all the files to data/raw directory and you can use the following script to generate a samples.tsv file automatically:

In [7]:
!(cd data/raw && ls *.raw > filelist.txt)
import pandas as pd
header_list = ["sample_name"]
df=pd.read_csv("data/raw/filelist.txt", names=header_list)
df["sample_name"]=df["sample_name"].replace(".raw", value="", regex=True)
df["comment"] = ""
df["MAPnumber"] = ""
df.to_csv("data/raw/samples.tsv", sep="\t")
df

sample_name comment MAPnumber
0    Epemicins                  
1  GermicidinA                  
2  GermicidinB

## Import libraries

In [8]:
from pyopenms import *
from pandas import DataFrame
import pandas as pd
import pyteomics
from pyteomics.openms import featurexml
import numpy as np
import os
import glob
from pyteomics import mztab

# Preprocessing step

## The first preprocessing function consists of five steps:

### 1) PrecursorCorrection (To the "highest intensity MS1 peak")

This algorithm is used directly after the file introduction, in order to correct any wrong MS1 precursor annotation - this means it will correct the precursos m/z in a given mass range (e.g. 0.1 Da) and then use the highest intensity peak as the precursor. We assume that, in the given mass window, the precursor with the hightest intensity was actually fragmented (top-n method), which is a method used by default in the Thermo Orbitrap instrument (Center for Biosustainability).

### OpenMS definition:

"Selection of the peak with the highest intensity as corrected precursor mass in a given mass range (e.g. precursor mass +/- 0.2 Da)

For each MS2 spectrum the corresponding MS1 spectrum is determined by using the rt information of the precursor. In the MS1, the peak with the highest intensity in a given mass range to the uncorrected precursor m/z is selected and used as corrected precursor m/z."

#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html//classOpenMS_1_1PrecursorCorrection.html#a8acf85ba8b9f249de0369bb083355982 

###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959

In [9]:
path= "results/interim"
isExist= os.path.exists(path)
if not isExist:
    os.mkdir(path)

input_original_files = glob.glob('results/*.mzML')
    
for filename in input_original_files:
    exp = MSExperiment()
    MzMLFile().load(filename, exp)
    exp.sortSpectra(True)
    delta_mzs= []
    mzs = []
    rts= []
    PrecursorCorrection.correctToHighestIntensityMS1Peak(exp, 100.0, True, delta_mzs, mzs, rts)
    mzmlfile_path = os.path.join(os.path.dirname(filename), "interim", "", 'precursorcorrected_' + os.path.basename(filename))
    MzMLFile().store(mzmlfile_path, exp)

In [10]:
path= "results/features/interim"
isExist= os.path.exists(path)
if not isExist:
    os.mkdir(path)

input_mzml_files = glob.glob('results/interim/*.mzML')

# 2) Mass trace detection

#A mass trace extraction method that gathers peaks similar in m/z and moving along retention time.
#Peaks of a MSExperiment are sorted by their intensity and stored in a list of potential chromatographic apex positions. Only peaks that are above the noise threshold (user-defined) are analyzed and only peaks that are n times above this minimal threshold are considered as apices. This saves computational resources and decreases the noise in the resulting output.
#Starting with these, mass traces are extended in- and decreasingly in retention time. During this extension phase, the centroid m/z is computed on-line as an intensity-weighted mean of peaks.
#The extension phase ends when either the frequency of gathered peaks drops below a threshold (min_sample_rate, see MassTraceDetection parameters) or when the number of missed scans exceeds a threshold (trace_termination_outliers, see MassTraceDetection parameters).
#Finally, only mass traces that pass a filter (a certain minimal and maximal length as well as having the minimal sample rate criterion fulfilled) get added to the result.

for filename in input_mzml_files:
    exp = MSExperiment()
    MzMLFile().load(filename, exp)
    exp.sortSpectra(True)
    mass_traces = []
    mtd = MassTraceDetection()
    mtd_par = mtd.getDefaults()
    mtd_par.setValue("mass_error_ppm", 10.0) 
    mtd_par.setValue("noise_threshold_int", 1.0e04)
    mtd.setParameters(mtd_par)
    mtd.run(exp, mass_traces, 0)

# 3) Elution peak detection

#Extracts chromatographic peaks from a mass trace.
#Mass traces may consist of several consecutively (partly overlapping) eluting peaks, e.g., stemming from (almost) isobaric compounds that are separated by retention time. Especially in metabolomics, isomeric compounds with exactly the same mass but different retentional behaviour may still be contained in the same mass trace.
#This method first applies smoothing on the mass trace's intensities, then detects local minima/maxima in order to separate the chromatographic peaks from each other. Detection of maxima is performed on the smoothed intensities and uses a fixed peak width (given as parameter chrom_fwhm) within which only a single maximum is expected. Currently smoothing is done using SavitzkyGolay smoothing with a second order polynomial and a frame length of the fixed peak width.
#Depending on the "width_filtering" parameters, mass traces are filtered by length in seconds ("fixed" filter) or by quantile.
#The output of the algorithm is a set of chromatographic peaks for each mass trace, i.e. a vector of split mass traces (see ElutionPeakDetection parameters).
#In general, a user would want to call the "detectPeaks" functions, potentially followed by the "filterByPeakWidth" function.
#This method in other words is "deconvolution".

    mass_traces_split = []
    mass_traces_final = []
    epd = ElutionPeakDetection()
    epd_par = epd.getDefaults()
    epd_par.setValue("width_filtering", "fixed")
    epd.setParameters(epd_par)
    epd.detectPeaks(mass_traces, mass_traces_split)
     
    if (epd.getParameters().getValue("width_filtering") == "auto"):
          epd.filterByPeakWidth(mass_traces_split, mass_traces_final)
    else:
          mass_traces_final = mass_traces_split

# 4) Feature detection

#FeatureFinderMetabo assembles metabolite features from singleton mass traces.
#Mass traces alone would allow for further analysis such as metabolite ID or statistical evaluation. However, in general, monoisotopic mass traces are accompanied by satellite C13 peaks and thus may render the analysis more difficult. FeatureFinderMetabo fulfills a further data reduction step by assembling compatible mass traces to metabolite features (that is, all mass traces originating from one metabolite). To this end, multiple metabolite hypotheses are formulated and scored according to how well differences in RT (optional), m/z or intensity ratios match to those of theoretical isotope patterns.
#If the raw data scans contain the scan polarity information, it is stored as meta value "scan_polarity" in the output file.
#Mass trace clustering can be done using either 13C distances or a linear model (Kenar et al) – see parameter 'ffm:mz_scoring_13C'. Generally, for lipidomics, use 13C, since lipids contain a lot of 13C. For general metabolites, the linear model is usually more appropriate. To decide what is better, the total number of features can be used as indirect measure.
#the lower(!) the better (since more mass traces are assembled into single features). Detailed information is stored in the featureXML output: it contains meta-values for each feature about the mass trace differences (inspectable via TOPPView). 
#By default, the linear model is used.
    
    feature_map_FFM = FeatureMap()
    feat_chrom = []
    ffm = FeatureFindingMetabo()
    ffm_par = ffm.getDefaults() 
    ffm_par.setValue("isotope_filtering_model", "none")
    ffm_par.setValue("remove_single_traces", "true")
    ffm_par.setValue("mz_scoring_by_elements", "false")
    ffm_par.setValue("report_convex_hulls", "true")
    ffm.setParameters(ffm_par)
    ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)
    feature_map_FFM.setUniqueIds()
    feature_map_FFM.setPrimaryMSRunPath([filename.encode()])  


# 5) Metabolite adduct decharger (MetaboliteFeatureDeconvolution)

#For each peak, this algorithm reconstructs neutral masses by enumerating all possible adducts with matching charge. 
#You can add the list of adduct files and database files for the algorithm to parse through.
#With SIRIUS, an algorithm that is later used, you are only able to use singly charged adducts so charges higher than 1 are filtered out.

    mfd = MetaboliteFeatureDeconvolution()
    mdf_par = mfd.getDefaults()
    mdf_par.setValue("potential_adducts",  [b"H:+:0.6",b"Na:+:0.2",b"NH4:+:0.1", b"H2O:-:0.1"])
    mdf_par.setValue("charge_min", 1, "Minimal possible charge")
    mdf_par.setValue("charge_max", 1, "Maximal possible charge")
    mdf_par.setValue("charge_span_max", 1)
    mdf_par.setValue("max_neutrals", 1)
    mfd.setParameters(mdf_par)
        
    feature_maps = FeatureMap()
    cons_map0 = ConsensusMap()
    cons_map1 = ConsensusMap()
    mfd.compute(feature_map_FFM, feature_maps, cons_map0, cons_map1)
    featurefile = os.path.join("results", "", "features", "", "interim", "", 'MFD_' + os.path.basename(filename)[19:-5] +".featureXML")
    FeatureXMLFile().store(featurefile, feature_maps)

Progress of 'mass trace detection':
-- done [took 0.93 s (CPU), 0.13 s (Wall)] -- 
Progress of 'elution peak detection':
-- done [took 0.35 s (CPU), 0.05 s (Wall)] -- 
Progress of 'assembling mass traces to features':
-- done [took 0.32 s (CPU), 0.04 s (Wall)] -- 
MassExplainer table size: 12
4611 spectra and 1 chromatograms stored.
3619 spectra and 1 chromatograms stored.
3625 spectra and 1 chromatograms stored.
Generating Masses with threshold: -6.90776 ...
done
0 of 2 valid net charge compomer results did not pass the feature charge constraints
Inferring edges raised edge count from 2 to 2
Found 2 putative edges (of 31) and avg hit-size of 1
Using solver 'coinor' ...
Optimal solution found!
 Branch and cut took 0.028533 seconds,  with objective value: 0.18.
ILP score is: 0.18
Agreeing charges: 4/4


Compomer: Da -41.0003; q_net 0; logP -3.91202[[ (H2O1Na1) --> () ]]
 Compomer: Da -36.0449; q_net 0; logP -4.60517[[ (H2O1H4N1) --> () ]]
 Compomer: Da -21.9819; q_net 0; logP -2.12026[[ (Na1) --> (H1) ]]
 Compomer: Da -19.0184; q_net 0; logP -2.81341[[ (H1H2O1) --> () ]]
 Compomer: Da -17.0265; q_net 0; logP -2.81341[[ (H4N1) --> (H1) ]]
 Compomer: Da -4.9554; q_net 0; logP -3.91202[[ (Na1) --> (H4N1) ]]
 Compomer: Da 4.9554; q_net 0; logP -3.91202[[ (H4N1) --> (Na1) ]]
 Compomer: Da 17.0265; q_net 0; logP -2.81341[[ (H1) --> (H4N1) ]]
 Compomer: Da 19.0184; q_net 0; logP -2.81341[[ () --> (H1H2O1) ]]
 Compomer: Da 21.9819; q_net 0; logP -2.12026[[ (H1) --> (Na1) ]]
 Compomer: Da 36.0449; q_net 0; logP -4.60517[[ () --> (H2O1H4N1) ]]
 Compomer: Da 41.0003; q_net 0; logP -3.91202[[ () --> (H2O1Na1) ]]
 

Progress of 'mass trace detection':
-- done [took 1.49 s (CPU), 0.53 s (Wall)] -- 
Progress of 'elution peak detection':
-- done [took 2.32 s (CPU), 0.37 s (Wall)] -- 
Progress of 'assembling mass traces to features':
-- done [took 4.47 s (CPU), 0.64 s (Wall)] -- 
MassExplainer table size: 12
Generating Masses with threshold: -6.90776 ...
done
0 of 107 valid net charge compomer results did not pass the feature charge constraints
Inferring edges raised edge count from 107 to 107
Found 107 putative edges (of 3855) and avg hit-size of 1
Using solver 'coinor' ...
Optimal solution found!
 Branch and cut took 0.023603 seconds,  with objective value: 4.56.
ILP score is: 4.56
Agreeing charges: 190/190
Progress of 'mass trace detection':
-- done [took 0.84 s (CPU), 0.14 s (Wall)] -- 
Progress of 'elution peak detection':
-- done [took 0.36 s (CPU), 0.05 s (Wall)] -- 
Progress of 'assembling mass traces to features':
-- done [took 0.38 s (CPU), 0.05 s (Wall)] -- 


Compomer: Da -41.0003; q_net 0; logP -3.91202[[ (H2O1Na1) --> () ]]
 Compomer: Da -36.0449; q_net 0; logP -4.60517[[ (H2O1H4N1) --> () ]]
 Compomer: Da -21.9819; q_net 0; logP -2.12026[[ (Na1) --> (H1) ]]
 Compomer: Da -19.0184; q_net 0; logP -2.81341[[ (H1H2O1) --> () ]]
 Compomer: Da -17.0265; q_net 0; logP -2.81341[[ (H4N1) --> (H1) ]]
 Compomer: Da -4.9554; q_net 0; logP -3.91202[[ (Na1) --> (H4N1) ]]
 Compomer: Da 4.9554; q_net 0; logP -3.91202[[ (H4N1) --> (Na1) ]]
 Compomer: Da 17.0265; q_net 0; logP -2.81341[[ (H1) --> (H4N1) ]]
 Compomer: Da 19.0184; q_net 0; logP -2.81341[[ () --> (H1H2O1) ]]
 Compomer: Da 21.9819; q_net 0; logP -2.12026[[ (H1) --> (Na1) ]]
 Compomer: Da 36.0449; q_net 0; logP -4.60517[[ () --> (H2O1H4N1) ]]
 Compomer: Da 41.0003; q_net 0; logP -3.91202[[ () --> (H2O1Na1) ]]
 Compomer: Da -41.0003; q_net 0; logP -3.91202[[ (H2O1Na1) --> () ]]
 Compomer: Da -36.0449; q_net 0; logP -4.60517[[ (H2O1H4N1) --> () ]]
 Compomer: Da -21.9819; q_net 0; logP -2.12026[[

MassExplainer table size: 12
Generating Masses with threshold: -6.90776 ...
done
0 of 7 valid net charge compomer results did not pass the feature charge constraints
Inferring edges raised edge count from 7 to 7
Found 7 putative edges (of 38) and avg hit-size of 1
Using solver 'coinor' ...
Optimal solution found!
 Branch and cut took 0.003425 seconds,  with objective value: 0.38.
ILP score is: 0.38
Agreeing charges: 10/10


### Convert the FeatureXML files to dataframes

In [11]:
from collections import defaultdict
from functools import reduce
from pathlib import Path
from time import perf_counter
import sys

from IPython.core.display import display
from pandas import CategoricalDtype
import numpy as np
from pyopenms import *
import pandas as pd
import os

common_meta_value_types = {
    b'label': 'U30',
    b'spectrum_index': 'i',
    b'score_fit': 'f',
    b'score_correlation': 'f',
    b'FWHM': 'f',
    b'spectrum_native_id': 'U30',
    b"num_of_masstraces" : "f"
}

class FeatureMapDF(FeatureMap):
    def __init__(self):
        super().__init__()
    
    # meta_values = None (default), 'all' or list of meta value names
    def get_df(self, meta_values = None):
        # get all possible meta value keys in a set
        if meta_values == 'all':
            meta_values = set()
            for f in self:
                mvs = []
                f.getKeys(mvs)
                for m in mvs:
                    meta_values.add(m)
        elif not meta_values: # if None, set to empty list
            meta_values = []
        
        def gen(fmap: FeatureMap, fun):
            for f in fmap:
                yield from fun(f, meta_values)

        def extract_meta_data(f: Feature, meta_values):
            pep = f.getPeptideIdentifications()  # type: list[PeptideIdentification]
            bb = f.getConvexHull().getBoundingBox2D()
                
            vals = [f.getMetaValue(m) if f.metaValueExists(m) else np.NA for m in meta_values]   # find some NA or None value for numpy
            
            if len(pep) != 0:
                hits = pep[0].getHits()

                if len(hits) != 0:
                    besthit = hits[0]  # type: PeptideHit
                    yield f.getUniqueId(), besthit.getSequence().toString(), f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue("num_of_masstraces"), f.getOverallQuality(), f.getIntensity()
                else:
                    yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue("num_of_masstraces"), f.getOverallQuality(), f.getIntensity()
            else:
                yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue("num_of_masstraces"), f.getOverallQuality(), f.getIntensity()

        cnt = self.size()

        mddtypes = [('id', np.dtype('uint64')), ('sequence', 'U200'), ('charge', 'i4'), ('RT', 'f'), ('mz', 'f'),
                    ('RTstart', 'f'), ('RTend', 'f'), ("num_of_masstraces", 'f'),
                    ('quality', 'f'), ('intensity', 'f')]
        
        for meta_value in meta_values:
            if meta_value in common_meta_value_types:
                mddtypes.append((meta_value.decode(), common_meta_value_types[meta_value]))
            else:
                mddtypes.append((meta_value.decode(), 'U50'))
        mdarr = np.fromiter(iter=gen(self, extract_meta_data), dtype=mddtypes, count=cnt)
        df= pd.DataFrame(mdarr).set_index('id').sort_values("mz").drop(columns= "sequence")
        #df= df[df["num_of_masstraces"]>=2]
        return df

In [12]:
input_feature_files = glob.glob('results/features/interim/*.featureXML')

for filename in input_feature_files:
    fmap = FeatureMapDF()
    FeatureXMLFile().load(filename, fmap)
    DF= fmap.get_df()
    feature_csv= os.path.join("results", "", "features", "", 'features_' + os.path.basename(filename)[23:-10] +"csv")
    DF.to_csv(feature_csv)
    display(DF)

charge          RT          mz     RTstart       RTend  \
id                                                                             
7828444526003282178        1  148.905807   79.029442   53.512493  153.915115   
5790822193567747154        1   54.525616   80.024490   52.503883  267.433807   
12103250118592697438       1  614.040466  111.519051  473.607147  659.460144   
6476063448694421522        1  301.385956  121.068192  296.368378  307.432556   
14036854744335096349       1  311.435425  121.068192  308.441589  375.975830   
6642006441343023006        1  658.436279  124.086975  601.664368  659.460144   
2631226070276202208        1   47.200829  127.039024   42.465252   52.503883   
17338785395939283382       1   56.553089  129.012909   52.503883   82.733200   
5513755335874708566        2   56.553089  137.001694   52.503883  370.922058   
547381735650245826         2   57.548325  137.476105   52.503883  216.132828   
16599045891870183703       1  658.436279  140.081665  632.584045  659.460144   
18279564239868844669       1   45.138130  141.958694   44.076504   52.503883   
5077808391661443930        2   56.553089  143.990234   52.503883   98.789200   
12006853981872375352       1   54.525616  144.489487   52.503883   72.632843   
492605022259058839         2   55.540855  144.987869   52.503883  354.796173   
13566907338101906037       1   47.200829  145.049530   42.465252   52.503883   
10485901738229864845       2   56.553089  145.985825   52.503883  268.436798   
714269606906917943         1   56.553089  147.984238   52.503883   64.547096   
11487335791239747973       1  511.282349  149.023468  474.631104  578.789368   
1152445925541224625        1   55.540855  150.951096   50.355537  442.184967   
8504423148768620179        2   55.540855  153.489212   52.503883  181.999863   
16937614398427051530       1   45.138130  158.961365   44.076504   51.429779   
5703527676895442697        2   54.525616  167.486649   52.503883  268.436798   
8172364719383552202        1  411.627319  171.149200  334.621185  442.184967   
10843342768381195178       1  511.282349  177.054749  508.118164  601.664368   
10840101699129665141       1  658.436279  179.092758  611.990967  659.460144   
14142219859237140692       1   53.512493  179.948151   52.503883   83.735931   
10303191250439308077       1  418.978302  183.101624  415.819977  441.125000   
16334676574646105181       1  415.819977  186.221649  197.048874  659.460144   
12244094377544316479       1  445.191559  197.117386  442.184967  659.460144   
13329455854415769547       1   47.200829  200.972382   44.076504   51.429779   
5898814370456702364        1  456.254395  207.149323  454.233795  462.348938   
18422087468178586152       1  406.428009  207.159073  404.372253  411.627319   
9782502892790774151        1   54.525616  212.954819   52.503883   74.663811   
4688076228965549470        1   56.553089  213.962662   52.503883  209.107208   
15809892714626394650       1   54.525616  217.014359   52.503883  442.184967   
10251381782329901386       1   54.525616  220.911057   52.503883   90.780823   
5445427694899417587        1   54.525616  223.943649   52.503883  442.184967   
1783545213949887123        1  493.370605  225.196304  490.241150  566.119507   
2764784098518548950        1  482.918152  228.196014  474.631104  514.469788   
12919264476909748298       1  586.107117  235.169357  574.603516  659.460144   
401333331107403434         1   54.525616  237.959686   52.503883  229.189896   
5570085132482620393        1   54.525616  245.967789   52.503883  386.097809   
3169913793095001305        1   54.525616  252.965530   52.503883  237.242645   
12636726805999279348       1   54.525616  256.959961   52.503883  389.124146   
11069558902728809626       1   54.525616  264.967926   52.503883  388.115997   
963353895330429588         1   54.525616  269.968292   52.503883  399.277985   
15904351843606998647       1   54.525616  273.962738   52.503883  432.659485   
12825080102515410184  

charge          RT          mz     RTstart       RTend  \
id                                                                             
7533231791879286860        1   53.185089   79.021217   51.165127  381.049072   
1915485981838167057        1  262.347748   80.024498   58.195900  419.430908   
15767548886940136318       1  262.347748   81.016991   63.226200  419.430908   
3414371255935680560        1  636.506348  111.519012  509.028076  659.194153   
8173649458821225888        1  296.508972  114.091362   90.442024  415.385468   
4585161104175024601        2   54.180222  114.986160   51.165127  162.835083   
1071191568748756725        1  311.570862  121.068192  308.549042  372.974731   
7972845349525779272        1  300.520172  121.068192  293.495026  309.564026   
13368136595716820518       2   54.180222  137.001694   45.045105  371.973236   
13591020934362218328       2   54.180222  137.476105   52.187241  222.212219   
11794809249556952260       2   54.180222  143.990234   52.187241  160.824997   
4567553093664634320        2   55.192169  144.987885   51.165127  364.929352   
3975509239172342202        1   49.110657  144.987930   45.045105   50.140896   
1561140909547581390        1   55.192169  145.985825   48.098415  319.634125   
6609795209173386345        1   58.195900  148.493744   52.187241   69.246971   
7234028246274782235        1   54.180222  150.951096   43.008492  506.876862   
7531339304116258534        2   55.192169  153.009979   52.187241  161.830978   
14751034771204165631       2   54.180222  154.993332   52.187241  163.834930   
16525594826740923621       1   55.192169  165.489471   52.187241   62.214966   
12674850098321046912       1   64.222931  165.489517   63.226200   89.432045   
5757143126172037092        1  418.423157  168.078140  416.408936  476.789185   
12333265125248530650       1  412.329132  171.149200  392.110931  434.673187   
17575077180014116509       1  658.163513  179.092758  620.078308  659.194153   
8822808628289013806        1  418.423157  183.101624  414.363129  659.194153   
11510354729325957249       1   58.195900  186.221634   42.461220  152.789993   
9003871463182089571        1   51.165127  197.117218   42.461220   60.188625   
6776000009256072689        1  406.245972  207.159088  404.223480  415.385468   
14955989571711251982       1   53.185089  212.954819   45.045105   81.367912   
17486769506543743262       1   55.192169  223.912003   52.187241  135.716980   
10860358900423552869       1   54.180222  223.943649   44.003857  563.960632   
1531372773260466756        1  492.883209  225.196182  490.723602  573.460022   
13460787753879252840       1  482.164612  228.195877  475.726196  503.651794   
9268600333021860384        1  585.995605  235.169296  423.437225  659.194153   
13923699278614003474       1   56.187649  237.959702   44.003857   91.444000   
4886908609525262654        1   54.180222  245.967789   51.165127  402.193665   
17908740997019243465       1   54.180222  252.965515   50.140896  191.076904   
14588058089510613206       1  409.289062  255.065231  407.266968  414.363129   
11706967964205344192       1   53.185089  256.959961   51.165127  408.278076   
8559557642666076520        1   49.110657  256.959991   44.003857   50.140896   
8959650122752827728        1   52.187241  257.959808   49.110657  279.383087   
17778284737602076773       1   54.180222  262.972534   49.110657  275.367035   
13407004761837739536       1   54.180222  269.968292   49.110657  411.315033   
15970604336759335573       1   54.180222  273.962738   51.165127  416.408936   
9593795525270639828        1  500.416565  274.274261  498.255707  511.177551   
3963672802648736484        1  486.443268  279.093475  448.262848  659.194153   
12920359819961185337       1  641.626770  284.910919  460.877411  659.194153   
8147696270070961921        1   54.180222  301.925903   53.185089   69.246971   
4268919152908008813        1   53.185089  301.957611   51.165127  473.597870   
9136209135308997509   

charge          RT           mz     RTstart       RTend  \
id                                                                              
14345265954344242653       1  329.607452   208.133240  326.792847  335.102264   
4904190476273846959        1  310.614807   208.133255  303.933105  315.405701   
3368833473880021387        1  379.585754   260.185669  362.698792  455.897400   
11190109598164828968       1  228.511551   261.123413  222.785919  235.216217   
8446623022672685115        1  316.367584   271.060028  302.927643  321.092896   
...                      ...         ...          ...         ...         ...   
12044417715849511886       1  350.433289  1603.895020  348.765778  355.269623   
2413724936652252017        1  342.791321  1604.396851  337.522614  352.043243   
10014946678412038621       2   47.941818  1639.049805   43.264481   61.428474   
9878374773279591185        1  173.429962  1646.474365  169.596756  181.128769   
3508833745391403603        2  105.360298  1775.552734  101.564934  115.742714   

                      num_of_masstraces   quality     intensity  
id                                                               
14345265954344242653                2.0  0.000180  3.212667e+07  
4904190476273846959                 2.0  0.000011  1.985221e+06  
3368833473880021387                 2.0  0.000258  4.589339e+07  
11190109598164828968                2.0  0.000025  4.301382e+06  
8446623022672685115                 2.0  0.000009  1.677704e+06  
...                                 ...       ...           ...  
12044417715849511886                3.0  0.001192  1.190649e+08  
2413724936652252017                 2.0  0.000371  4.679359e+07  
10014946678412038621                2.0  0.000003  2.294459e+05  
9878374773279591185                 3.0  0.000006  5.192035e+05  
3508833745391403603                 6.0  0.000049  1.904661e+06  

[2459 rows x 8 columns]

### 6) PrecursorCorrection (To the "nearest feature”)

This algorithm is used after feature detection, adduct grouping and even identification via accurate mass search. It basically allows the precursor correction on MS2 level. 

Which means that if there are MS2 spectra in my feature space which have been measured in isotope traces, it “corrects” the MS2 spectrum annotation to the monoisotopic trace. That is why you have a high mass deviation 100 pm, but 0.0 rt tolerance. So it basically corrects the MS2 to the feature centroid that can be found/mapped by SIRIUS::preprocessing later on.

Wrong assignment of the mono-isotopic mass for precursors are assumed:

1. if precursor_mz matches the mz of a non-monoisotopic feature mass trace and 
2. in the case that believe_charge is true: if feature_charge matches the precursor_charge In the case of wrong mono-isotopic assignment several options for correction are available: keep_original will create a copy of the precursor and tandem spectrum for the new mono-isotopic mass trace and retain the original one. all_matching_features does this not for only the closest feature but all features in a question.

#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1PrecursorCorrection.html

###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959

#### We need to preprocess again without convex hulls this time, otherwise the results from SIRIUS are incorrect

### (ii) SIRIUS Adapter
    The SIRIUS function is optional and includes the SIRIUS Adapter Algorithm from the Boecher lab. 


    The algorithm generates formula prediction from scores calculated from 1) MS2 fragmentation scores (ppm error + intensity) and 2) MS1 isotopic pattern scores.


    It can only compute feautures that are singly charged. There is also a timeout for compounds (compound timeout so that it doesn't compute for longer than 100 seconds per feature, which normally happens with larger molecules).


    -sirius:compound_timeout <number>                    
    Maximal computation time in seconds for a single compound. 0 for an infinite amount of time. (default: '100' min: '0')

        
    This algorith can help in data dereplication and analysis for direct library search. 

#### (iii) CSI:FingerID

The CSI_fingerID function is another algorithm from the Boecher lab, just like SIRIUS adapter and is using the formula predictions from SIRIUS, to search in structural libraries and predict the structure of each formula

If you replace: 
    
    out_csifingerid = os.path.join("results", "", "interim", "", "CSI", "", 'structures_' + os.path.basename(filename) +".mzTab")

with an empty string:

    out_csifingerid = ""
        
CSI:FingerID will be ignored and will not be computed.

#### Documentation: https://boecker-lab.github.io/docs.sirius.github.io/

##### Citation: Kai Dührkop, Huibin Shen, Marvin Meusel, Juho Rousu, and Sebastian Böcker, Searching molecular structure databases with tandem mass spectra using CSI:FingerID, PNAS October 13, 2015 112 (41) 12580-12585, https://doi.org/10.1073/pnas.1509788112

In [ ]:
path1= "results/SIRIUS/interim"
isExist= os.path.exists(path1)
if not isExist:
    os.mkdir(path1)

path2= "results/CSI/interim"
isExist= os.path.exists(path2)
if not isExist:
    os.mkdir(path2)

for filename in input_mzml_files:
    exp = MSExperiment()
    MzMLFile().load(filename, exp)
    exp.sortSpectra(True)
    PrecursorCorrection.correctToNearestFeature(feature_map_DEC, exp, 0.0, 100.0, True, False, False, False, 3, 0)

    mass_traces = []
    mtd = MassTraceDetection()
    mtd_par = mtd.getDefaults()
    mtd_par.setValue("mass_error_ppm", 10.0) 
    mtd_par.setValue("noise_threshold_int", 1.0e04)
    mtd.setParameters(mtd_par)
    mtd.run(exp, mass_traces, 0)

    mass_traces_split = []
    mass_traces_final = []
    epd = ElutionPeakDetection()
    epd_par = epd.getDefaults()
    epd_par.setValue("width_filtering", "fixed")
    epd.setParameters(epd_par)
    epd.detectPeaks(mass_traces, mass_traces_split)
        
    if (epd.getParameters().getValue("width_filtering") == "auto"):
        epd.filterByPeakWidth(mass_traces_split, mass_traces_final)
    else:
        mass_traces_final = mass_traces_split

    feature_map_FFM = FeatureMap()
    feat_chrom = []
    ffm = FeatureFindingMetabo()
    ffm_par = ffm.getDefaults() 
    ffm_par.setValue("isotope_filtering_model", "none")
    ffm_par.setValue("remove_single_traces", "false")
    ffm_par.setValue("mz_scoring_by_elements", "false")
    ffm.setParameters(ffm_par)
    ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)
    feature_map_FFM.setUniqueIds()
    feature_map_FFM.setPrimaryMSRunPath([filename.encode()])

    mfd = MetaboliteFeatureDeconvolution()
    mdf_par = mfd.getDefaults()
    mdf_par.setValue("potential_adducts",  [b"H:+:0.6",b"Na:+:0.2",b"NH4:+:0.1", b"H2O:-:0.1"])
    mdf_par.setValue("charge_min", 1, "Minimal possible charge")
    mdf_par.setValue("charge_max", 1, "Maximal possible charge")
    mdf_par.setValue("charge_span_max", 1)
    mdf_par.setValue("max_neutrals", 1)
    mfd.setParameters(mdf_par)
    feature_map_DEC = FeatureMap()
    cons_map0 = ConsensusMap()
    cons_map1 = ConsensusMap()
    mfd.compute(feature_map_FFM, feature_map_DEC, cons_map0, cons_map1)
    
    featureinfo = os.path.join("results", "", "SIRIUS", "", "interim", "", 'MFD_noconvexhulls_' + os.path.basename(filename)[:-5] + ".featureXML")
    FeatureXMLFile().store(featureinfo, feature_map_DEC)

    sirius_algo = SiriusAdapterAlgorithm()
    sirius_algo_par = sirius_algo.getDefaults()
    sirius_algo_par.setValue("preprocessing:filter_by_num_masstraces", 2) #Number of mass traces each feature has to have to be included
    sirius_algo_par.setValue("preprocessing:feature_only", "true") #Uses the feature information from in_featureinfo to reduce the search space to MS2
    sirius_algo_par.setValue("sirius:profile", "orbitrap")
    sirius_algo_par.setValue("sirius:db", "none")
    sirius_algo_par.setValue("sirius:ions_considered", "[M+H]+, [M-H2O+H]+, [M+Na]+, [M+NH4]+")
    sirius_algo_par.setValue("sirius:elements_enforced", "CHN[15]OS[4]") 
    sirius_algo_par.setValue("project:processors", 2)
    sirius_algo_par.setValue("fingerid:db", "BIO")
    sirius_algo.setParameters(sirius_algo_par)
        

    fm_info = FeatureMapping_FeatureMappingInfo()
    feature_mapping = FeatureMapping_FeatureToMs2Indices() 
    sirius_algo.preprocessingSirius(featureinfo,
                                    exp,
                                    fm_info,
                                    feature_mapping)
    sirius_algo.logFeatureSpectraNumber(featureinfo, 
                                        feature_mapping,
                                        exp)
    msfile = SiriusMSFile()
    debug_level = 3
    sirius_tmp = SiriusTemporaryFileSystemObjects(debug_level)
    siriusstring= String(sirius_tmp.getTmpMsFile())
    feature_only = sirius_algo.isFeatureOnly()
    isotope_pattern_iterations = sirius_algo.getIsotopePatternIterations()
    no_mt_info = sirius_algo.isNoMasstraceInfoIsotopePattern()
    compound_info = []
    msfile.store(exp,
                String(sirius_tmp.getTmpMsFile()),
                feature_mapping, 
                feature_only,
                isotope_pattern_iterations, 
                no_mt_info, 
                compound_info)
    
    out_csifingerid = os.path.join("results", "", "CSI", "", "interim", "", 'structures_' + os.path.basename(filename)[19:-5] +".mzTab")
    executable= "resources/Sirius/sirius.app/Contents/MacOS/sirius"
    subdirs = sirius_algo.callSiriusQProcess(String(sirius_tmp.getTmpMsFile()),
                                            String(sirius_tmp.getTmpOutDir()),
                                            String(executable),
                                            String(out_csifingerid),
                                            False)
    candidates = sirius_algo.getNumberOfSiriusCandidates()
    sirius_result = MzTab()
    siriusfile = MzTabFile()
    SiriusMzTabWriter.read(subdirs,
                        filename,
                        candidates,
                        sirius_result)
    
    sirius_file= os.path.join("results", "", "SIRIUS", "", "interim", "",'formulas_' + os.path.basename(filename)[19:-5] +".mzTab")
    siriusfile.store(sirius_file, sirius_result)

#### Convert mzTab files to dataframes

In [7]:
input_SIRIUS_files = glob.glob('results/SIRIUS/interim/*.mzTab')
for filename in input_SIRIUS_files:
    sirius=  pyteomics.mztab.MzTab(filename, encoding='UTF8', table_format='df')
    sirius.metadata
    df= sirius.small_molecule_table
    SIRIUS_DF= df.drop(columns= ["identifier", "smiles", "inchi_key", "description", "calc_mass_to_charge", "charge", "taxid", "species","database", "database_version", "spectra_ref", "search_engine", "modifications"])
    sirius_DF_file= os.path.join("results", "", "SIRIUS", "", os.path.basename(filename)[-6] +".csv")
    SIRIUS_DF.to_csv(sirius_DF_file)

In [8]:
# Example formula predictions dataframe
import pandas as pd
SIRIUS_DF= pd.read_csv("./results/SIRIUS/formulas_Epemicins.csv", index_col= "Unnamed: 0")
SIRIUS_DF=SIRIUS_DF[SIRIUS_DF["opt_global_explainedIntensity"] >= 0.6] #opt_global_explainedIntensity should be higher than 0.8 or 0.9 even for reliable results
SIRIUS_DF= SIRIUS_DF.sort_values(by= "exp_mass_to_charge")
SIRIUS_DF

chemical_formula  exp_mass_to_charge  retention_time  \
2782        C7H16N3OS          208.133234      329.607460   
2781        C10H12N3O          208.133234      329.607460   
3687        C14H13NO3          261.123425      228.511545   
3688       C12H11N4O2          261.123425      228.511545   
3689        C7H13N7OS          261.123425      228.511545   
...               ...                 ...             ...   
551   C45H22N12O24S10         1451.827148      415.963700   
555    C31H26N10O40S8         1451.827148      415.963700   
554   C38H38N10O20S15         1451.827148      415.963700   
552   C30H42N12O22S16         1451.827148      415.963700   
553   C37H26N14O26S11         1451.827148      415.963700   

      best_search_engine_score[1]  best_search_engine_score[2]  \
2782                    13.604441                    13.604441   
2781                    15.230887                    14.193324   
3687                    46.917919                    46.549750   
3688                    39.257591                    39.257591   
3689                    34.713566                    34.713566   
...                           ...                          ...   
551                     29.538484                    29.538484   
555                     29.538484                    29.538484   
554                     29.538484                    29.538484   
552                     29.538484                    29.538484   
553                     29.538484                    29.538484   

      best_search_engine_score[3] opt_global_adduct  \
2782                     0.000000     [M + H4 + N]+   
2781                     1.037563     [M + H4 + N]+   
3687                     0.368169     [M + H4 + N]+   
3688                     0.000000     [M + H4 + N]+   
3689                     0.000000     [M + H4 + N]+   
...                           ...               ...   
551                      0.000000     [M + H4 + N]+   
555                      0.000000     [M + H4 + N]+   
554                      0.000000     [M + H4 + N]+   
552                      0.000000     [M + H4 + N]+   
553                      0.000000     [M + H4 + N]+   

     opt_gobal_precursorFormula  opt_global_rank  opt_global_explainedPeaks  \
2782                  C7H20N3OS                3                          4   
2781                  C10H16N3O                2                          4   
3687                  C14H17NO3                1                         11   
3688                 C12H15N4O2                2                         14   
3689                  C7H17N7OS                3                         13   
...                         ...              ...                        ...   
551             C45H26N12O24S10                1                         12   
555              C31H30N10O40S8                5                         12   
554             C38H42N10O20S15                4                         12   
552             C30H46N12O22S16                2                         12   
553             C37H30N14O26S11                3                         12   

      opt_global_explainedIntensity  \
2782                       0.935599   
2781                       0.935599   
3687                       0.833473   
3688                       0.996021   
3689                       0.982498   
...                             ...   
551                        0.756276   
555                        0.756276   
554                        0.756276   
552                        0.756276   
553                        0.756276   

      opt_global_median_mass_error_fragment_peaks_ppm  \
2782                                         6.190684   
2781                                         6.881382   
3687                                        -1.250390   
3688                                         5.349610   
3689                                         7.144945   
...                                               ...   
5

In [20]:
input_CSI_files = glob.glob('results/CSI/interim/*.mzTab')
for filename in input_CSI_files:
    CSI=  pyteomics.mztab.MzTab(filename, encoding='UTF8', table_format='df')
    CSI.metadata
    DF= CSI.small_molecule_table
    DF= DF.drop(columns= ["calc_mass_to_charge", "charge", "taxid", "species","database", "database_version", "spectra_ref", "search_engine", "modifications"])
    csifingerID_file= os.path.join("results", "", "CSI", "" + os.path.basename(filename)[:-6]+ ".csv")
    DF.to_csv(csifingerID_file)

In [21]:
#example formula predictions table for Epemicins
csifingerID= pd.read_csv("./results/CSI/structures_Epemicins.csv",index_col= "Unnamed: 0")
csifingerID

identifier chemical_formula  \
0     14852549|14852550|25053655|24971283|53652498|5...        C25H45NO5   
1                                      9889418|11059239        C25H45NO5   
2                                     71464556|71464557        C25H45NO5   
3                                               9824494        C25H45NO5   
4                                               3049003        C25H45NO5   
...                                                 ...              ...   
2572  288197|2754113|7056395|7056399|11868690|118686...         C19H28O5   
2573                                                 ""         C19H28O5   
2574                             354385|494326|23425264         C19H28O5   
2575                                                 ""         C19H28O5   
2576                                           11717027         C19H28O5   

                                                 smiles       inchi_key  \
0          CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)CNC=O  ANPULBVRORHAGO   
1            CCCCCCCCCCC1C(OC1=O)CC(CCCCCCC)OC(=O)CNC=O  RQFZMIXBLJEUGS   
2     CCCCCC=CCC=CCCCCCC(CC(=O)OC(CC(=O)O)C[N+](C)(C...  WQYXCASYXUFNSI   
3        CCCCCCCC(CC1C(C(=O)O1)CCCCCCCC(C)C)OC(=O)CNC=O  SMEWLQMERHEMJK   
4     CCCCCCC(C)(CCN1C(C(=O)C(C1=O)(C)C)CCCCCCC(=O)O...  OHQKFFPFFFUNJW   
...                                                 ...             ...   
2572        CC12CCC3C(C1CCC2=O)CCC(C3(C)CC(=O)O)CC(=O)O  CEUQHWSIXGFNGI   
2573   CC(=O)C(=O)C1(CCC2C(C(=O)CCC2(C1CC(=O)O)C)(C)C)C  QWLGLRAGSFVBDI   
2574    CC(=O)OC1CC2CC3(C(CC(C3C2(C1=C)O)(C)C)OC(=O)C)C  CLIPYJQRTQNGAM   
2575    CC(=O)OC1CC2CC3(CC(C(C3C2(C1=C)O)(C)C)OC(=O)C)C  KEVCQSXXJWIVSR   
2576             CCCCCCCC(C(C(C#CC#CC(C=C)O)O)OC(=O)C)O  TYSPCPWTSAFCRA   

                                            description  exp_mass_to_charge  \
0                              Orlistat derivative, 15g          457.363953   
1                                                   NaN          457.363953   
2                            3-hydroxylinoleylcarnitine          457.363953   
3                                                   NaN          457.363953   
4     Ethyl 4,4-dimethyl-3,5-dioxo-1-(3-hydroxy-3-me...          457.363953   
...                                                 ...                 ...   
2572                                                NaN          337.199054   
2573                                                NaN          337.199054   
2574  .delta.-(sup 9,8.beta., 10.alpha.-triol, 3,8-d...          337.199054   
2575                                                NaN          337.199054   
2576  [(8R,9R,10R)-3,8,10-trihydroxyheptadec-1-en-4,...          337.199054   

      retention_time  best_search_engine_score[1]  opt_global_rank  \
0         327.701974                  -167.839042                1   
1         327.701974                  -168.290573                2   
2         327.701974                  -170.979093                3   
3         327.701974                  -171.211550                4   
4         327.701974                  -250.749662                5   
...              ...                          ...              ...   
2572      465.541832                  -155.792453                1   
2573      465.541832                  -168.039526                2   
2574      465.541832                  -203.768295                3   
2575      465.541832                  -205.873518                4   
2576      465.541832                  -210.247162                5   

      opt_global_compoundId  opt_global_compoundScanNumber  \
0                      1571                           1572   
1                      1571                           1572   
2                      1571                           1572   
3                      1571                           1572   
4                      1571                           1572   
...                     ...                            ...   
2572   

#### Explanation of columns
##### mz= mass-to-charge ratio (m/z)
##### RT= retention time (sec)
##### intensity = intensity of the feature (AU-arbitrary units)
##### FWHM= Full Width of the peak at Half its Maximum height
##### num_of_masstraces	= number of mass traces detected (single mass traces are excluded). This is relevant to the isotopic pattern
##### isotope_distances = distance in mz between the isotopes (jumps of app. 1 is important to confirm that this is a real feature) 

#### Create a metadata csv file for GNPS from the samples.tsv file 

In [13]:
path= "results/GNPSexport/interim"
isExist= os.path.exists(path)
if not isExist:
    os.mkdir(path)

df= pd.read_csv("data/raw/samples.tsv", sep= "\t", index_col= "Unnamed: 0")
metadata= df.rename(columns= {"sample_name": "filename", "comment": "ATTRIBUTE_comment", "MAPnumber": "ATTRIBUTE_MAPnumber"})
metadata["filename"]= metadata["filename"].astype(str) +".mzml"
metadata['ATTRIBUTE_MAPnumber'] = np.arange(len(metadata))
metadata["ATTRIBUTE_comment"]= "MAP" + metadata["ATTRIBUTE_MAPnumber"].astype(str)
metadata= metadata.drop(columns= "ATTRIBUTE_MAPnumber")
metadata['ATTRIBUTE_genomeID']=metadata['filename'].str.extract(r'(NBC_?\d*)')
metadata['ATTRIBUTE_genomeIDMDNA']=metadata['filename'].str.extract(r'(MDNA_WGS_?\d*)')
metadata['ATTRIBUTE_genomeID']=metadata['ATTRIBUTE_genomeID'].fillna(metadata['ATTRIBUTE_genomeIDMDNA'])
metadata=metadata.drop(columns="ATTRIBUTE_genomeIDMDNA")
#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'NBC', value= 'NBC_', regex= True)
#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'MDNAWGS', value= 'MDNA_WGS_', regex= True)
metadata.to_csv("results/GNPSexport/metadata.tsv", sep='\t')
metadata

filename ATTRIBUTE_comment  ATTRIBUTE_genomeID
0    Epemicins.mzml              MAP0                 NaN
1  GermicidinA.mzml              MAP1                 NaN
2  GermicidinB.mzml              MAP2                 NaN

#### MapAlignerPoseClustering is used to perform a linear retention time alignment, basically correct for linear shifts in retention time. 
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_MapAlignerPoseClustering.html

In [41]:
# get in index of feature map with highest number of features in feature map list
path= "results/consensus/interim/"
isExist= os.path.exists(path)
if not isExist:
    os.mkdir(path)

input_feature_files = glob.glob('results/features/interim/*.featureXML')
feature_maps = []
for featurexml_file in input_feature_files:
    fmap = FeatureMap()
    FeatureXMLFile().load(featurexml_file, fmap)
    feature_maps.append(fmap)
    
ref_index = [i[0] for i in sorted(enumerate([fm.size() for fm in feature_maps]), key=lambda x:x[1])][-1]

aligner = MapAlignmentAlgorithmPoseClustering()
aligner_par= aligner.getDefaults()
aligner_par.setValue("max_num_peaks_considered", -1)
aligner.setParameters(aligner_par)
aligner.setReference(feature_maps[ref_index])

for feature_map in feature_maps[:ref_index] + feature_maps[ref_index+1:]:
    trafo = TransformationDescription()
    aligner.align(feature_map, trafo)
    transformer = MapAlignmentTransformer()
    transformer.transformRetentionTimes(feature_map, trafo, True) # store original RT as meta value
    
for feature_map in feature_maps:    
    feature_file = os.path.join("results", "", "consensus", "", "interim", "", 'MapAligned_' + os.path.basename(feature_map.getMetaValue('spectra_data')[0].decode())[19:-5] +".featureXML")
    trafo_file= os.path.join("results", "", "consensus", "", "interim", "", 'MapAligned_' + os.path.basename(feature_map.getMetaValue('spectra_data')[0].decode())[19:-5] +".trafoXML")
    FeatureXMLFile().store(feature_file, feature_map)
    TransformationXMLFile().store(trafo_file, trafo)

#### Introduce the features to a protein identification file (idXML)   
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_IDMapper.html

In [34]:
import pyopenms as pms

use_centroid_rt= False
use_centroid_mz= True
protein_ids = []
peptide_ids= []

mapper = pms.IDMapper()

for filename in input_mzml_files:
    exp = MSExperiment()
    MzMLFile().load(filename, exp)

    for fmap in feature_maps: 
        if os.path.basename(fmap.getMetaValue('spectra_data')[0].decode()) == os.path.basename(filename):

            mapper.annotate(fmap, peptide_ids, protein_ids, use_centroid_rt, use_centroid_mz, exp)
            
            for index_protein, protein in enumerate(fmap.getProteinIdentifications()):
                protein.setIdentifier(str.encode(str(index_protein)))
                #print(protein.getIdentifier)
            for index_feature, feature in enumerate(fmap):
                for index_pep, pep in enumerate(feature.getPeptideIdentifications()):
                    pep.setIdentifier(str.encode(str(index_feature)+'_'+str(index_pep)))
                    #print(pep.getIdentifier())
        featureidx_file = os.path.join("results", "", "consensus", "", "interim", "", 'IDMapper_' + os.path.basename(fmap.getMetaValue('spectra_data')[0].decode())[19:-5] +".featureXML")
        FeatureXMLFile().store(featureidx_file, fmap)

Unassigned peptides: 0
Peptides assigned to exactly one feature: 0
Peptides assigned to multiple features: 0
Unassigned and unidentified precursors: 3015
Unidentified precursor assigned to exactly one feature: 0
Unidentified precursor assigned to multiple features: 0
Feature annotation with identifications:
    no ID: 1321
    single ID: 0
    multiple IDs (identical): 0
    multiple IDs (divergent): 0


Unassigned peptides: 0
Peptides assigned to exactly one feature: 0
Peptides assigned to multiple features: 0
Unassigned and unidentified precursors: 3587
Unidentified precursor assigned to exactly one feature: 0
Unidentified precursor assigned to multiple features: 0
Feature annotation with identifications:
    no ID: 14069
    single ID: 0
    multiple IDs (identical): 0
    multiple IDs (divergent): 0


Unassigned peptides: 0
Peptides assigned to exactly one feature: 0
Peptides assigned to multiple features: 0
Unassigned and unidentified precursors: 3020
Unidentified precursor assign

#### The FeatureLinkerUnlabeledKD is used to aggregate the feature information (from single files) into a ConsensusFeature, linking features from different files together, which have a smiliar m/z and rt (no MS2 data).
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FeatureLinkerUnlabeledKD.html

In [42]:
class ConsensusMapDF(ConsensusMap):
    def __init__(self):
        super().__init__()

    def get_intensity_df(self):
        labelfree = self.getExperimentType() == "label-free"
        filemeta = self.getColumnHeaders()  # type: dict[int, ColumnHeader]
        labels = list(set([header.label for header in
                           filemeta.values()]))  # TODO could be more efficient. Do we require same channels in all files?
        files = list(set([header.filename for header in filemeta.values()]))
        label_to_idx = {k: v for v, k in enumerate(labels)}
        file_to_idx = {k: v for v, k in enumerate(files)}

        def gen(cmap: ConsensusMap, fun):
            for f in cmap:
                yield from fun(f)

        if not labelfree:
            # TODO write two functions for LF and labelled. One has only one channel, the other has only one file per CF
            def extractRowBlocksChannelWideFileLong(f: ConsensusFeature):
                subfeatures = f.getFeatureList()  # type: list[FeatureHandle]
                filerows = defaultdict(lambda: [0] * len(labels))  # TODO use numpy array?
                for fh in subfeatures:
                    header = filemeta[fh.getMapIndex()]
                    row = filerows[header.filename]
                    row[label_to_idx[header.label]] = fh.getIntensity()
                return (f.getUniqueId(), filerows)

            def extractRowsChannelWideFileLong(f: ConsensusFeature):
                uniqueid, rowdict = extractRowBlocksChannelWideFileLong(f)
                for file, row in rowdict.items():
                    row.append(file)
                    yield tuple([uniqueid] + row)

            if len(labels) == 1:
                labels[0] = "intensity"
            dtypes = [('id', np.dtype('uint64'))] + list(zip(labels, ['f'] * len(labels)))
            dtypes.append(('file', 'U300'))
            # For TMT we know that every feature can only be from one file, since feature = PSM
            #cnt = 0
            #for f in self:
            #    cnt += f.size()

            intyarr = np.fromiter(iter=gen(self, extractRowsChannelWideFileLong), dtype=dtypes, count=self.size())
            return pd.DataFrame(intyarr).set_index('id')
        else:
            # Specialized for LabelFree which has to have only one channel
            def extractRowBlocksChannelLongFileWideLF(f: ConsensusFeature):
                subfeatures = f.getFeatureList()  # type: list[FeatureHandle]
                row = [0.] * len(files)  # TODO use numpy array?
                for fh in subfeatures:
                    header = filemeta[fh.getMapIndex()]
                    row[file_to_idx[header.filename]] = fh.getIntensity()
                yield tuple([f.getUniqueId()] + row)

            dtypes = [('id', np.dtype('uint64'))] + list(zip(files, ['f'] * len(files)))
            # cnt = self.size()*len(files) # TODO for this to work, we would need to fill with NAs for CFs that do not go over all files
            cnt = self.size()

            intyarr = np.fromiter(iter=gen(self, extractRowBlocksChannelLongFileWideLF), dtype=dtypes, count=cnt)
            return pd.DataFrame(intyarr).set_index('id')

    def get_metadata_df(self):
        def gen(cmap: ConsensusMap, fun):
            for f in cmap:
                yield from fun(f)

        def extractMetaData(f: ConsensusFeature):
            # subfeatures = f.getFeatureList()  # type: list[FeatureHandle]
            pep = f.getPeptideIdentifications()  # type: list[PeptideIdentification]
            if len(pep) != 0:
                hits = pep[0].getHits()
                if len(hits) != 0:
                    besthit = hits[0]  # type: PeptideHit
                    # TODO what else
                    yield f.getUniqueId(), besthit.getSequence().toString(), f.getCharge(), f.getRT(), f.getMZ(), f.getQuality()
                else:
                    yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), f.getQuality()
            else:
                yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), f.getQuality()

        cnt = self.size()

        mddtypes = [('id', np.dtype('uint64')), ('sequence', 'U200'), ('charge', 'i4'), ('RT', 'f'), ('mz', 'f'),
                    ('quality', 'f')]
        mdarr = np.fromiter(iter=gen(self, extractMetaData), dtype=mddtypes, count=cnt)
        return pd.DataFrame(mdarr).set_index('id')

In [43]:
feature_grouper = FeatureGroupingAlgorithmKD()

consensus_map = ConsensusMapDF()
file_descriptions = consensus_map.getColumnHeaders()

for i, feature_map in enumerate(feature_maps):
    file_description = file_descriptions.get(i, ColumnHeader())
    file_description.filename = feature_map.getMetaValue('spectra_data')[0].decode()
    file_description.size = feature_map.size()
    file_description.unique_id = feature_map.getUniqueId()
    file_descriptions[i] = file_description

    consensus_map.setColumnHeaders(file_descriptions)
    feature_grouper.group(feature_maps, consensus_map)


    Consensus_file= os.path.join("results", "", "consensus", "","interim", "", 'consensus' + ".consensusXML")
    ConsensusXMLFile().store(Consensus_file, consensus_map)


# get intensities as a DataFrame
intensities = consensus_map.get_intensity_df()

# get meta data as DataFrame
meta_data = consensus_map.get_metadata_df()[['RT', 'mz', 'quality']]

# you can concatenate these two for a "result" DataFrame
result = pd.concat([meta_data, intensities], axis=1)

# if you don't need labeled index, remove it (and/or save with index = False)
result.reset_index(drop=True, inplace=True)

# store as tsv file
result.to_csv('results/consensus/Consensus.tsv', sep = '\t', index = False)


                Progress of 'computing RT transformations':

                -- done [took 0.09 s (CPU), 0.10 s (Wall)] -- 

                Progress of 'linking features':

                -- done [took 0.20 s (CPU), 0.19 s (Wall)] -- 
ConsensusMap contains 15390 invalid references to maps:
  wrong id=1 (occurred 1321x)
  wrong id=2 (occurred 14069x)


                Progress of 'computing RT transformations':

                -- done [took 0.11 s (CPU), 0.10 s (Wall)] -- 

                Progress of 'linking features':

                -- done [took 0.19 s (CPU), 0.19 s (Wall)] -- 
ConsensusMap contains 14069 invalid references to maps:
  wrong id=2 (occurred 14069x)


                Progress of 'computing RT transformations':

                -- done [took 0.10 s (CPU), 0.10 s (Wall)] -- 

                Progress of 'linking features':

                -- done [took 0.18 s (CPU), 0.18 s (Wall)] -- 


The ConsensusXML file contains invalid maps or references thereof. Please fix the file or notify the maintainer of this tool if you did not provide a consensusXML file! Note that this warning will be a fatal error in the next version of OpenMS!
The ConsensusXML file contains invalid maps or references thereof. Please fix the file or notify the maintainer of this tool if you did not provide a consensusXML file! Note that this warning will be a fatal error in the next version of OpenMS!


#### Filter out the features that do not have an MS2 pattern
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FileFilter.html

    rule FileFilter:
        input:
            "results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML"
        output:
            "results/Consensus/filtered.consensusXML"
        shell:
            """
            resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in {input} -out {output} 
            """


In [ ]:
!resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML -out results/Consensus/filtered.consensusXML


#### GNPS_export creates an mgf file with only the MS2 information of all files (introduce mzml files with spaces between them)
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_GNPSExport.html

    rule GNPS_export:
        input:
            "results/Consensus/filtered.consensusXML",
            expand("results/{samples}/interim/precursorcorrected_{samples}.mzML", samples=SAMPLES)
        output:
            "results/GNPSexport/MSMS.mgf" 
        shell:
            """
            resources/OpenMS-2.7.0/bin/GNPSExport -in_cm {input[0]} -in_mzml {input[1]} -out {output} 
            """

In [15]:
input_mzml_files=glob.glob("results/interim/*.mzML")
# load ms data from mzML file into MSExperiment
mzml= []
for mzml_file in input_mzml_files:
    spectra = MSExperiment()
    MzMLFile().load(mzml_file, spectra)
    mzml.append(spectra)

In [ ]:
!resources/OpenMS-2.7.0/bin/GNPSExport -ini resources/GNPSExport.ini -in_cm results/consensus/interim/filtered.consensusXML -in_mzml {mzml} -out results/GNPSexport/MSMS.mgf

#### export the consensusXML file to a txt file for GNPS
Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_TextExporter.html
    
    rule txt_export:
        input:
            "results/Consensus/filtered.consensusXML"
        output:
            "results/GNPSexport/FeatureQuantificationTable.txt" 
        shell:
            """
            resources/OpenMS-2.7.0/bin/TextExporter -in {input} -out {output}
            """

In [ ]:
!resources/OpenMS-2.7.0/bin/TextExporter -in results/Consensus/filtered.consensusXML -out results/GNPSexport/FeatureQuantificationTable.txt

TextExporter took 0.09 s (wall), 0.09 s (CPU), 0.01 s (system), 0.08 s (user); Peak Memory Usage: 19 MB.


### Workflow with FeatureFinderMetaboIdentCompound

In [2]:
import csv
# read tsv file and create list of FeatureFinderMetaboIdentCompound
def metaboTableFromFile(path_to_library_file):
    metaboTable = []
    with open(path_to_library_file, 'r') as tsv_file:
        tsv_reader = csv.reader(tsv_file, delimiter="\t")
        next(tsv_reader) # skip header
        for row in tsv_reader:
            metaboTable.append(FeatureFinderMetaboIdentCompound(
                row[0], # name
                row[1], # sum formula
                float(row[2]), # mass
                [int(charge) for charge in row[3].split(',')], # charges
                [float(rt) for rt in row[4].split(',')], # RTs
                [float(rt_range) for rt_range in row[5].split(',')], # RT ranges
                [float(iso_distrib) for iso_distrib in row[6].split(',')] # isotope distributions
            ))
    return metaboTable

In [3]:
import glob
from pyopenms import *

input_mzml_files=glob.glob("results/interim/*.mzML")
# load ms data from mzML file into MSExperiment
for mzml_file in input_mzml_files:
    spectra = MSExperiment()
    MzMLFile().load(mzml_file, spectra)

    # create FeatureFinderAlgorithmMetaboIdent and assign ms data
    ff = FeatureFinderAlgorithmMetaboIdent()
    ff.setMSData(spectra)

    # read library generate a metabo table with compounds
    metabo_table = metaboTableFromFile('MetaboliteIdentification.tsv')

    # FeatureMap to store results
    fm = FeatureMap()

    # edit some parameters
    params = ff.getParameters()
    params[b'extract:mz_window'] = 10.0 
    #params[b'extract:rt_window'] = 2.0 
    params[b'detect:peak_width'] = 60.0 
    ff.setParameters(params)

    # run the FeatureFinderMetaboIdent with the metabo_table and store results in fm
    ff.run(metabo_table, fm)

    # save FeatureMap to file
    ff_file = os.path.join("results", "", "FFMI", "", "interim", "", 'FFMI_' + os.path.basename(mzml_file)[19:-5] +".featureXML")
    FeatureXMLFile().store(ff_file, fm) 
